packages/cohort_sdk/src/cohort.rs

```diff
@@ -436,79 +473,39 @@ impl Cohort {
         }
     }
 
-    async fn await_for_snapshot<S>(&self, state_provider: &S, previous_conflict: Option<u64>, timeout: Duration) -> Result<CapturedState, SnapshotPollErrorType>
+    async fn await_for_snapshot<F, Fut>(
+        &self,
+        get_state_callback_fn: &F,
+        previous_conflict: Option<u64>,
+        timeout: Duration,
+    ) -> Result<CertificationCandidateCallbackResponse, SnapshotPollErrorType>
     where
-        S: ItemStateProvider,
+        F: Fn() -> Fut,
+        Fut: Future<Output = Result<CertificationCandidateCallbackResponse, String>>,
     {
-        match previous_conflict {
-            None => state_provider.get_state().await.map_err(|reason| SnapshotPollErrorType::FetchError { reason }),
-            Some(conflict) => {
-                let mut delay_controller = DelayController::new(self.config.backoff_on_conflict.min_ms, self.config.backoff_on_conflict.max_ms);
-                let poll_started_at = Instant::now();
-                loop {
-                    let result_local_state = state_provider.get_state().await;
-                    match result_local_state {
-                        Err(reason) => return Err(SnapshotPollErrorType::FetchError { reason }),
-                        Ok(current_state) => {
-                            if current_state.abort_reason.is_some() {
-                                break Ok(current_state);
-                            } else if current_state.snapshot_version < conflict {
-                                // not safe yet
-                                let waited = poll_started_at.elapsed();
-                                if waited >= timeout {
-                                    return Err(SnapshotPollErrorType::Timeout { waited });
-                                }
-                                delay_controller.sleep().await;
-                                continue;
-                            } else {
-                                break Ok(current_state);
-                            }
-                        }
+        let conflict = previous_conflict.unwrap_or(0);
```

Comment:
I was wondering how did you flatten the Optionals, here is the answer, a bit of cheating  with 0 :)

Reply:
that is a safe cheating :)